### ISSUE TITLE
Is that the right CV-folds?

### ISSUE THREAD

**Issue by jmq19950824** [rasbt/mlxtend#521]
When I set the cv-folds=5 in the Stacked classifier and deploy verbose=1 in the StackingCVClassifier function.  The training processes are shown as following:
Training and fitting fold 1 of 2...
Training and fitting fold 2 of 2...

I am confused cause the result seems there are only 2 folds total in my CV method. However, I do set cv=5 in my stacking models( please see the following code):

 Model_Stack=GridSearchCV(sclf,params_grid_Stack,scoring='accuracy',
                                 cv=5,n_jobs=corenum,refit=True).fit(X_train.values,y_train.values)

Can anybody tell me what's going on ?

**Comment by qiagu** [comment]
There are some known issues with gridsearch. I will have some time to work on the `StackingCVClassifier` and other estimators next week.  The issue you mentioned might be cleared after that.

**Comment by rasbt** [comment]
Hm, I think it was fine before we added the `safe_indexing` approach recently. Not sure what happened there, because the unit tests worked all well and reproduced the results compared to before using `safe_indexing`. We should maybe add an explicit check for ensuring that the number of folds matches the number provided -- my feeling is that it should be fine but rather a print error

**Comment by jmq19950824** [comment]
> Hm, I think it was fine before we added the `safe_indexing` approach recently. Not sure what happened there, because the unit tests worked all well and reproduced the results compared to before using `safe_indexing`. We should maybe add an explicit check for ensuring that the number of folds matches the number provided -- my feeling is that it should be fine but rather a print error

Thanks. As you said, I have tried various cv-folds and figure that whether different hyper-parameters will be selected,i.e., If the above situation is only caused by the print error, the optimal hyper-parameters will be different based on various cv-folds. However, I get the same selection results. So there may be some errors in this packages and this issue should be corrected.

**Comment by rasbt** [comment]
Hm, are you sure you set the number of folds to 5 in the StackingCVClassifier? I tried several values and couldn't reproduce this bug. Also, which version of mlxtend are you using? I tried the following with mlxtend 0.15 which seems to work as expected:

```python
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB 
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
from mlxtend.classifier import StackingCVClassifier
from sklearn import datasets

import mlxtend
print(mlxtend.__version__)

iris = datasets.load_iris()
X, y = iris.data[:, 1:3], iris.target


clf1 = KNeighborsClassifier()
clf2 = RandomForestClassifier()
clf3 = GaussianNB()
lr = LogisticRegression()

sclf = StackingCVClassifier(classifiers=[clf1, clf2, clf3], 
                            meta_classifier=lr,
                            cv=4,
                            verbose=1)

params = {'kneighborsclassifier__n_neighbors': [1, 5],
          'randomforestclassifier__n_estimators': [10, 50],
          'meta-logisticregression__C': [0.1, 10.0]}

grid = GridSearchCV(estimator=sclf, 
                    param_grid=params, 
                    cv=5,
                    refit=True)
grid.fit(X, y)

cv_keys = ('mean_test_score', 'std_test_score', 'params')

for r, _ in enumerate(grid.cv_results_['mean_test_score']):
    print("%0.3f +/- %0.2f %r"
          % (grid.cv_results_[cv_keys[0]][r],
             grid.cv_results_[cv_keys[1]][r] / 2.0,
             grid.cv_results_[cv_keys[2]][r]))

print('Best parameters: %s' % grid.best_params_)
print('Accuracy: %.2f' % grid.best_score_)
```

![Screen Shot 2019-04-20 at 8 12 21 PM](https://user-images.githubusercontent.com/5618407/56464148-cdce0f80-63a8-11e9-9ace-9259ed4a60aa.png)

**Comment by jmq19950824** [comment]
![image](https://user-images.githubusercontent.com/39890732/56464482-f0017500-641c-11e9-81d4-91190cbb89e5.png)

**Comment by jmq19950824** [comment]
@rasbt Thanks a lot! It seems the problem has been solved. The error (actually not a error) is caused by the code StackingCVClassifier(classifiers=[clf1, clf2, clf3],  meta_classifier=lr, verbose=1)
Since I don't set the cv parameters, the default param is 2-folds and thus the bug like errors has appeard. I tried your codes and everything is ok (and my version is also 0.15)
![image](https://user-images.githubusercontent.com/39890732/56464489-158e7e80-641d-11e9-9652-7e807ab01ce8.png)

**Comment by jmq19950824** [comment]
@rasbt In addition, I would like to purpose some question about the mlxtend packages (and thank again for your kindness answer):
1. notice that your codes
sclf = StackingCVClassifier(classifiers=[clf1, clf2, clf3], 
                            meta_classifier=lr,
                            cv=4,
                            verbose=1)
GridSearchCV(estimator=sclf, 
                    param_grid=params, 
                    cv=5,
                    refit=True)
where the cv-folds seems not the same (4 and 5 ,respectively). Should we set the same folds number in both the StackingCVClassifier function and GridSearchCV function?

2. In your references (https://rasbt.github.io/mlxtend/user_guide/classifier/StackingCVClassifier/#references)
the Stacking classifiers' params are set as follows:
params = {'kneighborsclassifier-1__n_neighbors': [1, 5],
          'kneighborsclassifier-2__n_neighbors': [1, 5],
          'randomforestclassifier__n_estimators': [10, 50],
          'meta-logisticregression__C': [0.1, 10.0]}

In the above case, although the Stacking result may be greatest since we can select a bunch of  base classifiers which can learning different parts from the data and stack to a meta-classifier, the computational cost will be huge when much more base classifiers added into such framework. For example, I want to deploy eight models for stacking and each models have 1 hyper-parameter. If I set the grid length to 3 and finally I have to train 3^8=6561 models to select the best meta-classifier. This will be impossible due to the large work.
So, is there some alternative methods we can use? For example, I use the best base classifiers which are selected separatly on training data (and we only have to train 3*8=24 models) to stack the meta-classifier. This may not be as good as your method cause the bast base classifiers might repeatively learn the training data.

**Comment by rasbt** [comment]
> where the cv-folds seems not the same (4 and 5 ,respectively). Should we set the same folds number in both the StackingCVClassifier function and GridSearchCV function?

You can think of this as nested cross-validation, where there is an inner and an outer cross-validation loop. In this example, in the outer loop (Gridsearch), 4-folds (+1 test fold) are being used as training set in each iteration, and these are then further divided into folds for training the Stacking classifier. Hence, the inner folds (stacking cv folds) are a subset of the outer training folds in GridSearch.

The number doesn't have to be the same and is somewhat arbitrary. It also depends on the size of the training set. Like with regular cross-validation, I'd say a number between 5-10 is reasonable.

> This will be impossible due to the large work. So, is there some alternative methods we can use?

This is the big downside of GridSearch, and in practice, RandomSearch with a given time budget is often use. So, instead of using GridSearchCV, I suggest you give RandomizedSearchCV a try: https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.RandomizedSearchCV.html

**Comment by jmq19950824** [comment]
> > where the cv-folds seems not the same (4 and 5 ,respectively). Should we set the same folds number in both the StackingCVClassifier function and GridSearchCV function?
> 
> You can think of this as nested cross-validation, where there is an inner and an outer cross-validation loop. In this example, in the outer loop (Gridsearch), 4-folds (+1 test fold) are being used as training set in each iteration, and these are then further divided into folds for training the Stacking classifier. Hence, the inner folds (stacking cv folds) are a subset of the outer training folds in GridSearch.
> 
> The number doesn't have to be the same and is somewhat arbitrary. It also depends on the size of the training set. Like with regular cross-validation, I'd say a number between 5-10 is reasonable.
> 
> > This will be impossible due to the large work. So, is there some alternative methods we can use?
> 
> This is the big downside of GridSearch, and in practice, RandomSearch with a given time budget is often use. So, instead of using GridSearchCV, I suggest you give RandomizedSearchCV a try: https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.RandomizedSearchCV.html

Thanks  for your advice!

**Comment by jmq19950824** [comment]
@rasbt 
@qiagu 
Actually, I am newer from R to Python. So I wonder that is there any possibility that we can stack some deep learning algorithms in this Stacking framework? (For example, RNN or CNN)

**Comment by rasbt** [comment]
Unfortunately, this is not possible with the current framework as it is geared towards a scikit-learn-like API. However, scikit-learn itself supports simple multi-layer perceptrons that could be used. As deep learning models are usually in a more flexible way, I would suggest doing the stacking "manually" in that case based on your framework of choice, e.g., PyTorch

**Comment by qiagu** [comment]
FYI, Keras supports sklearn APIs, see https://keras.io/scikit-learn-api/. Besides, I recently hacked Keras model and made new APIs for 'better' compatibility. If you're interested, please have a look at https://github.com/ohsu-comp-bio/Galaxy-ML/pull/1

**Comment by jmq19950824** [comment]
@rasbt 
@qiagu 
Thank all of you for your kind advice! The above suggestion will be a good start for my studies with Python and I will try to Stacking some complicated models (e.g., LigheGBM , CatBoost or Deep Learning models). Now this issue could be closed if necessary.

**Comment by rasbt** [comment]
Oh that's very interesting to know. I am more of a PyTorch user and didn't know about that compatibility layer in Keras. May come in handy one day! In any case, glad that there is no looming issue and happy to close this thread then.

**Comment by qiagu** [comment]
@rasbt I am told that PyTorch is more popular nowadays, especially for ML proficients, whereas Keras is easier for beginners. I just started to learn deep learning, will switch to PyTorch some time later, I think.

**Comment by rasbt** [comment]
Yeah that sounds about right. You can think of Keras more something like scikit-learn whereas PyTorch is more like NumPy. However, it is maybe more like NumPy some features from Keras such as pre-defined layers. The reason why I mainly use PyTorch is that it's easier for doing "new" things (trying new architectures, loss functions, etc.) whereas Keras is probably better for applications (where the method development is not the main focus).

**Comment by qiagu** [comment]
Now I can image what PyTorch is like. Thanks,  @rasbt
